### PR TITLE
feat: Add `reportUsedIgnorePattern` option to `no-unused-vars` rule

### PR DESCRIPTION
**Pull request by Pearce-Ropion**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

- [x] Documentation update
- [ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
- [ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
- [x] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
- [ ] Add autofix to a rule
- [ ] Add a CLI option
- [ ] Add something to the core
- [ ] Other, please explain:

**What rule do you want to change?**

`no-unused-vars`

**What change do you want to make (place an "X" next to just one item)?**

- [x] Generate more warnings
- [ ] Generate fewer warnings
- [ ] Implement autofix
- [ ] Implement suggestions

**How will the change be implemented (place an "X" next to just one item)?**

- [x] A new option
- [ ] A new default behavior
- [ ] Other

**Please provide some example code that this change will affect:**

Example 1:
```js
/* eslint no-unused-vars: ["error", { "varsIgnorePattern": "^_" }] */

const _a = 5;
const _b = _a + 5
```

Example 2:
```js
/* eslint no-unused-vars: ["error", { "argsIgnorePattern": "^_" }] */

function foo(_a) {
  return _a + 5
}
```

Example 3:
```js
/* eslint no-unused-vars: ["error", { "destructuredArrayIgnorePattern": "^_" }] */

const [ a, _b ] = items;
console.log(a + _b);
```

Example 4:
```js
/* eslint no-unused-vars: ["error", { "caughtErrorsIgnorePattern": "^_" }] */

try {}
catch(_err) {
  console.error(_err);
}
```

**What does the rule currently do for this code?**

In all examples, no warnings are reported.

**What will the rule do after it's changed?**

When the new option `reportUsedIgnorePattern` is enabled, each of the examples above will report; indicating that a variable that has been marked as an ignored unused variable is being used.

#### What changes did you make? (Give an overview)

This change adds a new option `reportUsedIgnorePattern` to the `no-unused-vars` rule. When enabled, the rule will report any instances of variables that match any of the valid ignore pattern options when that variable is actually being used. 

Resolves: #17568

#### Is there anything you'd like reviewers to focus on?

When finding all references in order to report on the correct nodes, it is also finding the variable declaration. So a question for the team would be which nodes should this option report on:
* Should all references including the declaration be reported?
* Should only the variable declaration be reported? 
* Should only references not including the declaration be reported

### CODE REVIEW

File: lib/rules/no-unused-vars.js
Comment (costinsin) [inline lib/rules/no-unused-vars.js:222]:
```suggestion
                switch (true) {
                    case defType === "CatchClause" && config.caughtErrorsIgnorePattern:
                        type = "args";
                        pattern = config.caughtErrorsIgnorePattern.toString();
                        break;
                    case defType === "Parameter" && config.argsIgnorePattern:
                        type = "args";
                        pattern = config.argsIgnorePattern.toString();
                        break;
                    case defType !== "Parameter" && config.varsIgnorePattern:
                        type = "vars";
                        pattern = config.varsIgnorePattern.toString();
                    case def.name.parent.type === "ArrayPattern" && config.destructuredArrayIgnorePattern:
                        type = "elements of array destructuring";
                        pattern = config.destructuredArrayIgnorePattern.toString();
                        break;
                }
```

Comment (Pearce-Ropion) [reply lib/rules/no-unused-vars.js:222]: This seems like an anti pattern.

Also, there are 2 other functions that are very similar to this one in this rule already so making this change would be inconsistent with how the rest of the rule is formatted.

Comment (costinsin) [reply lib/rules/no-unused-vars.js:222]: This is clearly not an anti pattern, as TypeScript 5.3 [just added](https://devblogs.microsoft.com/typescript/announcing-typescript-5-3-beta/#switch-true-narrowing#switch-true-narrowing) type narrowing for this pattern. I feel it resembles the powerful `match` present usually in functional languages.

This being said, I agree with you that we should keep the same code style and not mix them. It was just a suggestion.

**Comment by mdjermanovic** [comment]
> * Should all references including the declaration be reported?
> * Should only the variable declaration be reported?
> * Should only references not including the declaration be reported

Would it make sense to report only on the reasons the variable is considered used, which are usually its read references?

This is how it currently works:

```js
/* eslint no-unused-vars: ["error", {
    varsIgnorePattern: "^_",
    reportUsedIgnorePattern: true
}] */

var _a; // not reported

_a = 1; // this is reported but doesn't seem relevant?

foo(_a); // this is also reported, ok

```

But this variable is considered used only because of the read reference so it might make sense to report like this:

```js
/* eslint no-unused-vars: ["error", {
    varsIgnorePattern: "^_",
    reportUsedIgnorePattern: true
}] */

var _a; // not reported

_a = 1; // not reported

foo(_a); // reported

```

File: lib/rules/no-unused-vars.js
Comment (mdjermanovic) [inline lib/rules/no-unused-vars.js:225]:
```suggestion
                    additional = `. Used ${type} must not match ${pattern}`;
```

**Comment by Pearce-Ropion** [comment]
> > * Should all references including the declaration be reported?
> > * Should only the variable declaration be reported?
> > * Should only references not including the declaration be reported
> 
> Would it make sense to report only on the reasons the variable is considered used, which are usually its read references?
> 
> This is how it currently works:
> 
> ```js
> /* eslint no-unused-vars: ["error", {
>     varsIgnorePattern: "^_",
>     reportUsedIgnorePattern: true
> }] */
> 
> var _a; // not reported
> 
> _a = 1; // this is reported but doesn't seem relevant?
> 
> foo(_a); // this is also reported, ok
> ```
> 
> But this variable is considered used only because of the read reference so it might make sense to report like this:
> 
> ```js
> /* eslint no-unused-vars: ["error", {
>     varsIgnorePattern: "^_",
>     reportUsedIgnorePattern: true
> }] */
> 
> var _a; // not reported
> 
> _a = 1; // not reported
> 
> foo(_a); // reported
> ```

That makes sense to me. However, I would also expect that a user shouldn't be allowed to write to an ignored unused variable since it should be unused.

File: lib/rules/no-unused-vars.js
Comment (mdjermanovic) [inline lib/rules/no-unused-vars.js:797]: Per this condition, `_a` wouldn't be reported at all in the following example, which seems wrong as it's used?

```js
/* eslint no-unused-vars: ["error", {
    varsIgnorePattern: "^_",
    reportUsedIgnorePattern: true
}]*/

var _a = 42;

foo(() => _a);

```

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

File: lib/rules/no-unused-vars.js
Comment (Pearce-Ropion) [reply lib/rules/no-unused-vars.js:797]: Do you think I should revert to just `ref => ref.from.variableScope === usedIgnoredVar.scope.variableScope`? Is there a better condition to use?

**Comment by Pearce-Ropion** [comment]
I changed it to skip all variable initializations but otherwise report on all references. I also doesn't check whether the variable scope matches anymore so it can handle variables defined in the outer scope which are used in an inner scope.

**Comment by Pearce-Ropion** [comment]
@mdjermanovic I think this is working as expected now. Can you think of any other cases that I could test for? Otherwise, I think we are good to go.

**Comment by mdjermanovic** [comment]
@Pearce-Ropion apologies for the delay in reviewing this PR.

> I changed it to skip all variable initializations but otherwise report on all references.

I'm not sure what the common use cases for the `varsIgnorePattern` option are, but if user has a code like the following where the rule doesn't report any errors:

```js
/*eslint no-unused-vars: ["error", { varsIgnorePattern: "^_", reportUsedIgnorePattern: true }]*/

var _foo;

_foo = 5;

```

and then they add more code that causes the variable to be considered used, so that the rule now reports it:

```js
/*eslint no-unused-vars: ["error", { varsIgnorePattern: "^_", reportUsedIgnorePattern: true }]*/

var _foo;

_foo = 5;

bar(_foo);

```

then, I think it makes the most sense to report only on the references where this rule considers the variable to be used: `_foo` in `bar(_foo)`, because `_foo = 5` has no effect on whether the variable is considered used, and it was valid before `bar(_foo)` was added.

Alternatively, we could report only the declaration assuming that the variable name is wrong because it was intended to be used (as opposed to reporting references, in which case the assumption is that the variable that wasn't intended to be used is mistakenly used).

Curious what others think about this.

**Comment by Pearce-Ropion** [comment]
No worries, eslint is a massive project and I don't want to burden anyone. 

> then, I think it makes the most sense to report only on the references where this rule considers the variable to be used: `_foo` in `bar(_foo)`, because `_foo = 5` has no effect on whether the variable is considered used, and it was valid before `bar(_foo)` was added.
 
> Alternatively, we could report only the declaration assuming that the variable name is wrong because it was intended to be used (as opposed to reporting references, in which case the assumption is that the variable that wasn't intended to be used is mistakenly used).

Using vscode's eslint extension as an example, on one hand it would be useful to see reported issues on the usages since that's what is considered invalid according to the new option. However, It would be more clear what the problem is and how to fix it at the declaration, so it'd probably be better to do it there.

**Comment by nzakas** [comment]
> then, I think it makes the most sense to report only on the references where this rule considers the variable to be used: _foo in bar(_foo), because _foo = 5 has no effect on whether the variable is considered used, and it was valid before bar(_foo) was added.

I think this makes the most sense. :+1:

@Pearce-Ropion sorry for the delay, we've been busy working on v9. Can you please rebase and make any further changes? Thanks!

**Comment by nzakas** [comment]
Oops, accidentally hit "close"

**Comment by Pearce-Ropion** [comment]
> Oops, accidentally hit "close"

lol, I was going to reopen once I pushed the changes.

**Comment by Pearce-Ropion** [comment]
This is still active. Waiting for a review.

File: lib/rules/no-unused-vars.js
Comment (Pearce-Ropion) [reply lib/rules/no-unused-vars.js:210]: As per https://github.com/eslint/eslint/pull/17662#discussion_r1365318959 we opted to leave the if statements as is to reduce the scope of the PR and to maintain consistency with other similar functions within the rule. If I make this change, should I also update the other functions that use a similar flow?

Comment (Pearce-Ropion) [reply lib/rules/no-unused-vars.js]: On second thought, given that this is the second time this suggestion has been made, I've gone ahead and refactored this and the other message getters in the rule for consistency to use `switch` statements.

Comment (nzakas) [reply lib/rules/no-unused-vars.js]: Please double-check who the person is who is leaving suggestions on a PR. Not everyone is a team member. :)

File: lib/rules/no-unused-vars.js
Comment (nzakas) [inline lib/rules/no-unused-vars.js:174]: Nice :+1:

**Review by nzakas — APPROVED** [review]
LGTM. Would like @mdjermanovic to re-review before merging.

**Comment by mdjermanovic** [comment]
In this case, the message seems misleading:

```js
/*eslint no-unused-vars: ["error", {
    "reportUsedIgnorePattern": true,
    "destructuredArrayIgnorePattern": "^_",
    "varsIgnorePattern": "[iI]gnored"
}]*/

let _x; // '_x' is marked as ignored but is used. Used vars must not match /[iI]gnored/u.

[_x] = arr;

foo(_x);

```

**Comment by Tanujkanti4441** [comment]
Hi @Pearce-Ropion, some changes are requested. you can check out.

File: tests/lib/rules/no-unused-vars.js
Comment (Pearce-Ropion) [reply tests/lib/rules/no-unused-vars.js]: Yep, nice catch. I also apparently got the error output correct on the invalid cases.

File: lib/rules/no-unused-vars.js
Comment (mdjermanovic) [inline lib/rules/no-unused-vars.js:107]:
```suggestion
            reportUsedIgnorePattern: false
```

To fix the lint error.

**Comment by mdjermanovic** [comment]
> In this case, the message seems misleading:
> 
> ```js
> /*eslint no-unused-vars: ["error", {
>     "reportUsedIgnorePattern": true,
>     "destructuredArrayIgnorePattern": "^_",
>     "varsIgnorePattern": "[iI]gnored"
> }]*/
> 
> let _x; // '_x' is marked as ignored but is used. Used vars must not match /[iI]gnored/u.
> 
> [_x] = arr;
> 
> foo(_x);
> ```

I'm still getting the same error, looks like this case still hasn't been addressed?

**Comment by mdjermanovic** [comment]
Here's a similar case that should also be fixed:

```js
/*eslint no-unused-vars: ["error", {
    "reportUsedIgnorePattern": true,
    "destructuredArrayIgnorePattern": "^_",
    "varsIgnorePattern": "[iI]gnored"
}]*/

const [ignored] = arr; // 'ignored' is marked as ignored but is used. Used elements of array destructuring must not match /^_/u.

foo(ignored);

```

The message should be: `'ignored' is marked as ignored but is used. Used vars must not match /[iI]gnored/u.`

File: lib/rules/no-unused-vars.js
Comment (mdjermanovic) [inline lib/rules/no-unused-vars.js:722]: Instead of collecting these vars and later figuring out why they were considered as used ignored, we could just report them right away with the data we have at this point:

```suggestion
                            if (config.reportUsedIgnorePattern && isUsedVariable(variable)) {
                                context.report({
                                    node: def.name,
                                    messageId: "usedIgnoredVar",
                                    data: {
                                        varName: variable.name,
                                        additional: `. Used elements of array destructuring must not match ${config.destructuredArrayIgnorePattern}`
                                    }
                                });
                            }
```

(the same for catch arguments, params and variables).

That would fix https://github.com/eslint/eslint/pull/17662#issuecomment-1910344522 and https://github.com/eslint/eslint/pull/17662#issuecomment-1959062359.

**Comment by nzakas** [comment]
@Pearce-Ropion are you still working on this?

**Comment by Pearce-Ropion** [comment]
> @Pearce-Ropion are you still working on this?

Yes, apologies. I've been slammed with work. I'll get the proposed suggestions in this weekend

**Comment by Pearce-Ropion** [comment]
Ok, I've applied @mdjermanovic's suggestions to report as soon as we detect the used ignored variable. I also explicitly added the test cases that @mdjermanovic used.

I've also updated the message data handling. I had already updated the other message data handlers as per previous comments, but I noticed that all 3 of the message data getters were doing basically the same thing. So I've added a `getVariableDescription` helper function to get the required info for the message data base on the type of the variable that is currently being reported.

**Review by nzakas — APPROVED** [review]
LGTM. Would like @mdjermanovic to review before merging.

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks for contributing!

**Pull request merged**
